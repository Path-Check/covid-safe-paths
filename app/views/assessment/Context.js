import { createContext } from 'react';

export const AnswersContext = createContext();
export const SurveyContext = createContext();
export const MetaContext = createContext();
