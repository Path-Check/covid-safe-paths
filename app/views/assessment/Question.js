import React, { useContext, useEffect, useMemo, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { SafeAreaView, ScrollView, StyleSheet, View } from 'react-native';

import { Typography } from '../../components/Typography';
import Colors from '../../constants/colors';
import { Button } from './Button';
import { AnswersContext } from './Context';
import { Option } from './Option';
import {
  QUESTION_TYPE_MULTI,
  SCREEN_TYPE_CHECKBOX,
  SCREEN_TYPE_DATE,
  SCREEN_TYPE_RADIO,
} from './constants';

/**
 * @typedef { import(".").SurveyQuestion } SurveyQuestion
 * @typedef { import(".").SurveyOption } SurveyOption
 * @typedef { import(".").SurveyAnswers } SurveyAnswers
 */

/** @type {React.FunctionComponent<{
 *   onNext: () => void;
 *   onChange: (value: { index: number, value: string }[]) => void;
 *   option: SurveyOption
 *   question: SurveyQuestion
 * }>} */
export const Question = ({ onNext, onChange, option, question }) => {
  const { t } = useTranslation();
  const answers = useContext(AnswersContext);
  const [selectedValues, setSelectedValues] = useState(
    answers[question.id] || [],
  );

  // Allow line breaks in the description
  const description = useMemo(() => {
    if (!question.question_description) return null;
    const elements = [];
    for (const line of question.question_description.split('\n')) {
      let l = line.trim();
      if (!l) continue;
      elements.push(
        <Typography testID='description' key={l} style={styles.description}>
          {l}
        </Typography>,
      );
    }
    return <View style={styles.descriptionWrapper}>{elements}</View>;
  }, [question.question_description]);

  const displayAsOption = [
    SCREEN_TYPE_CHECKBOX,
    SCREEN_TYPE_RADIO,
    SCREEN_TYPE_DATE,
  ].includes(question.screen_type);

  const options =
    displayAsOption &&
    option.values.map((option, index) => (
      <Option
        answer={selectedValues.find((v) => v.index === index)}
        index={index}
        key={option.value}
        onSelect={(value) => onSelectHandler(value, index)}
        option={option}
        isSelected={selectedValues.some((v) => v.index === index)}
        type={question.screen_type}
      />
    ));

  /** @type {(value: string, index: number) => void} */
  const onSelectHandler = (value, index) => {
    if (question.question_type === QUESTION_TYPE_MULTI) {
      return setSelectedValues((values) => {
        // TODO: Better way to filter single value questions?
        const singleValueQuestions = [
          'Choose not to answer',
          'None of the above',
          'None',
        ];
        // this looks for an existing value inside the selected values array
        // which indicates user unselected the value
        const unselectedValue = values.some((v) => v.index === index);
        const currentValue = option.values[index];
        // this handles deselect for values in multi question type
        // that should act as a single value type questions
        // basically when any other selection is made this question gets unselected
        const unselectSingleValueQuestion = values.filter((v) => {
          return !singleValueQuestions.includes(option.values[v.index].label);
        });

        if (singleValueQuestions.includes(currentValue.label)) {
          // this logic handles single value questions in multi question type environment
          return unselectedValue
            ? unselectSingleValueQuestion
            : [{ index, value }];
        } else {
          // this logic handles the multi value questions selection
          return unselectedValue
            ? unselectSingleValueQuestion.filter((v) => v.index !== index)
            : [...unselectSingleValueQuestion, { index, value }];
        }
      });
    }
    return setSelectedValues([{ index, value }]);
  };

  useEffect(() => {
    onChange(selectedValues);
  }, [selectedValues, onChange]);

  return (
    <SafeAreaView style={styles.container}>
      <View style={styles.header}>
        <Typography use='headline2'>{question.question_text}</Typography>
      </View>
      <ScrollView style={styles.scrollView}>
        <View style={styles.scrollViewContent}>
          {description}
          {options}
        </View>
      </ScrollView>
      <View style={styles.footer}>
        <Button
          disabled={!selectedValues.length}
          onPress={onNext}
          title={t('assessment.next')}
        />
      </View>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: Colors.SECONDARY_10,
  },
  header: {
    paddingHorizontal: 20,
    marginVertical: 40,
  },
  scrollView: {
    flex: 1,
  },
  scrollViewContent: {
    paddingHorizontal: 20,
    paddingVertical: 20,
  },
  descriptionWrapper: {
    marginBottom: 20,
  },
  footer: {
    padding: 20,
  },
});
