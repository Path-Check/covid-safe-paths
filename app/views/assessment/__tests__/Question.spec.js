import { fireEvent, render } from '@testing-library/react-native';
import React from 'react';
import { I18nextProvider } from 'react-i18next';

import i18n from '../../../locales/languages';
import { AnswersContext } from '../Context';
import { Question } from '../Question';
import {
  QUESTION_TYPE_MULTI,
  QUESTION_TYPE_TEXT,
  SCREEN_TYPE_CHECKBOX,
  SCREEN_TYPE_RADIO,
} from '../constants';

test('base', () => {
  const { asJSON } = render(
    <Question
      onChange={jest.fn()}
      onNext={jest.fn()}
      option={option}
      question={question}
    />,
    { wrapper: Wrapper },
  );
  expect(asJSON()).toMatchSnapshot();
});

test('supports line breaks in the description', () => {
  const { queryAllByTestId } = render(
    <Question
      onChange={jest.fn()}
      onNext={jest.fn()}
      option={option}
      question={{ ...question, question_description: 'Hello\nWorld' }}
    />,
    { wrapper: Wrapper },
  );
  expect(queryAllByTestId('description')).toHaveLength(2);
});

test('QUESTION_TYPE_TEXT', () => {
  let onChange = jest.fn();
  const { queryAllByTestId } = render(
    <Question
      onChange={onChange}
      onNext={jest.fn()}
      option={option}
      question={question}
    />,
    { wrapper: Wrapper },
  );
  fireEvent.press(queryAllByTestId('option')[0]);
  expect(onChange).toHaveBeenCalledWith([{ index: 0, value: '0' }]);
  fireEvent.press(queryAllByTestId('option')[1]);
  expect(onChange).toHaveBeenCalledWith([{ index: 1, value: '1' }]);
});

test('QUESTION_TYPE_MULTI', () => {
  let onChange = jest.fn();
  const { queryAllByTestId } = render(
    <Question
      onChange={onChange}
      onNext={jest.fn()}
      option={option}
      question={{
        question_type: QUESTION_TYPE_MULTI,
        screen_type: SCREEN_TYPE_CHECKBOX,
      }}
    />,
    { wrapper: Wrapper },
  );
  fireEvent.press(queryAllByTestId('option')[0]);
  expect(onChange).toHaveBeenCalledWith([{ index: 0, value: '0' }]);
  fireEvent.press(queryAllByTestId('option')[1]);
  expect(onChange).toHaveBeenCalledWith([
    { index: 0, value: '0' },
    { index: 1, value: '1' },
  ]);
  fireEvent.press(queryAllByTestId('option')[0]);
  expect(onChange).toHaveBeenCalledWith([{ index: 1, value: '1' }]);
});

function Wrapper({ children }) {
  return (
    <I18nextProvider i18n={i18n}>
      <AnswersContext.Provider value={{}}>{children}</AnswersContext.Provider>
    </I18nextProvider>
  );
}

const question = {
  option_key: 'option_1',
  question_key: '1',
  question_text: 'What is the answer?',
  question_type: QUESTION_TYPE_TEXT,
  required: false,
  screen_type: SCREEN_TYPE_RADIO,
};

const option = {
  key: 'option_1',
  values: [
    {
      label: 'A',
      value: '0',
    },
    {
      label: 'B',
      value: '1',
    },
  ],
};
