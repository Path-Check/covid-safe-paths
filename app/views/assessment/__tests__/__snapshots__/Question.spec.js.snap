// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`base 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "backgroundColor": "#f8f8ff",
        "flex": 1,
      }
    }
  >
    <RCTScrollView
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View>
        <View
          style={
            Object {
              "marginTop": 16,
              "paddingHorizontal": 20,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#1f2c9b",
                  "fontFamily": "IBMPlexSans-Bold",
                  "fontSize": 28,
                  "fontWeight": "500",
                  "lineHeight": 32,
                },
                Object {
                  "writingDirection": "ltr",
                },
                undefined,
              ]
            }
          >
            What is the answer?
          </Text>
        </View>
        <View
          style={
            Object {
              "paddingHorizontal": 20,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#2e2e2e",
                  "fontFamily": "IBMPlexSans",
                  "fontSize": 17,
                  "lineHeight": 24,
                },
                Object {
                  "writingDirection": "ltr",
                },
                Object {
                  "color": "#4754C5",
                  "fontFamily": "IBMPlexSans",
                  "fontSize": 17,
                  "lineHeight": 16,
                  "marginTop": 28,
                },
              ]
            }
          >
            Select one
          </Text>
          <View
            style={
              Object {
                "marginTop": 20,
              }
            }
          >
            <View
              accessible={true}
              focusable={true}
              isTVSelectable={true}
              style={
                Object {
                  "opacity": 1,
                }
              }
              testID="option"
            >
              <View
                style={
                  Array [
                    Object {
                      "backgroundColor": "#ffffff",
                      "borderColor": "#d3d3d3",
                      "borderRadius": 8,
                      "borderWidth": 2,
                      "marginBottom": 16,
                      "paddingHorizontal": 20,
                      "paddingVertical": 20,
                    },
                    false,
                  ]
                }
              >
                <View
                  style={
                    Object {
                      "alignItems": "flex-start",
                      "flexDirection": "row",
                    }
                  }
                >
                  <View
                    style={
                      Array [
                        Object {
                          "alignItems": "center",
                          "borderColor": "#999999",
                          "borderRadius": 40,
                          "borderWidth": 2,
                          "height": 24,
                          "justifyContent": "center",
                          "marginRight": 24,
                          "marginTop": 2,
                          "width": 24,
                        },
                        false,
                      ]
                    }
                  />
                  <Text
                    style={
                      Array [
                        Object {
                          "color": "#2e2e2e",
                          "fontFamily": "IBMPlexSans",
                          "fontSize": 17,
                          "lineHeight": 24,
                        },
                        Object {
                          "writingDirection": "ltr",
                        },
                        Object {
                          "color": "#2e2e2e",
                          "flex": 1,
                          "flexWrap": "wrap",
                          "fontFamily": "IBMPlexSans-Bold",
                          "fontSize": 19,
                          "fontWeight": "500",
                          "lineHeight": 24,
                        },
                      ]
                    }
                    testID="label"
                  >
                    A
                  </Text>
                </View>
              </View>
            </View>
            <View
              accessible={true}
              focusable={true}
              isTVSelectable={true}
              style={
                Object {
                  "opacity": 1,
                }
              }
              testID="option"
            >
              <View
                style={
                  Array [
                    Object {
                      "backgroundColor": "#ffffff",
                      "borderColor": "#d3d3d3",
                      "borderRadius": 8,
                      "borderWidth": 2,
                      "marginBottom": 16,
                      "paddingHorizontal": 20,
                      "paddingVertical": 20,
                    },
                    false,
                  ]
                }
              >
                <View
                  style={
                    Object {
                      "alignItems": "flex-start",
                      "flexDirection": "row",
                    }
                  }
                >
                  <View
                    style={
                      Array [
                        Object {
                          "alignItems": "center",
                          "borderColor": "#999999",
                          "borderRadius": 40,
                          "borderWidth": 2,
                          "height": 24,
                          "justifyContent": "center",
                          "marginRight": 24,
                          "marginTop": 2,
                          "width": 24,
                        },
                        false,
                      ]
                    }
                  />
                  <Text
                    style={
                      Array [
                        Object {
                          "color": "#2e2e2e",
                          "fontFamily": "IBMPlexSans",
                          "fontSize": 17,
                          "lineHeight": 24,
                        },
                        Object {
                          "writingDirection": "ltr",
                        },
                        Object {
                          "color": "#2e2e2e",
                          "flex": 1,
                          "flexWrap": "wrap",
                          "fontFamily": "IBMPlexSans-Bold",
                          "fontSize": 19,
                          "fontWeight": "500",
                          "lineHeight": 24,
                        },
                      ]
                    }
                    testID="label"
                  >
                    B
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
    <View
      style={
        Object {
          "padding": 20,
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="assessment-button"
      >
        <View
          style={
            Array [
              Object {
                "borderRadius": 10,
                "paddingHorizontal": 14,
                "paddingVertical": 22,
              },
              undefined,
              Object {
                "backgroundColor": "#e5e7fa",
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#2e2e2e",
                  "fontFamily": "IBMPlexSans",
                  "fontSize": 17,
                  "lineHeight": 24,
                },
                Object {
                  "writingDirection": "ltr",
                },
                Array [
                  Object {
                    "color": "#f8f8f8",
                    "textAlign": "center",
                  },
                  undefined,
                  Object {
                    "color": "#a5affb",
                  },
                ],
              ]
            }
          >
            Next
          </Text>
        </View>
      </View>
    </View>
  </RCTSafeAreaView>
</View>
`;
